### ISSUE TITLE
sharing multiple images or a set?

### ISSUE THREAD

**Issue by comotion** [perkeep/perkeep#1057]
sharing multiple items (or a set) of images from the ui. I was hoping the share url would present as a gallery the same way I see those items.
So I create a set by selecting them and hitting "Create new set with items"
Then I select the new set and hit Share, and get a share url. If I follow the share url, I get:
```
{"camliVersion": 1,
  "authType": "haveref",
  "camliSigner": "sha1-b7873b590283448c24ace1a7220ad787dbe76889",
  "camliType": "claim",
  "claimDate": "2018-02-14T10:16:14.943Z",
  "claimType": "share",
  "target": "sha1-575a906c4ea56bed90b4daf2d57e193d1150d6d5",
  "transitive": true
,"camliSig":"wsBcBAABCAAQBQKIbgkACRDCplLFYPcJngAAXOkIAAsqraN4H8ehbcyc9sJ5EfXkDwhxo/ChG6bXKDHoeLwHortlSiQDnltKTLxFQz3YjuwwoqHHoNJtl9infVzqGtBpyHpO/zE+J2snJEehYo2LhaYdE3OirHySSSaZT7aMJlIRPAolXwvkvBFpV7TUDhUCtcJgCudYXZaBdRnsg8MCHcygf/TyBZuJOHu7OhAPoaJojX7ilJJtwiZj67xUS2T7UvpE9NCtNAX/ShqkOKOXgBjboEr9+E0HhQfvlvvt5qu6C6U06mUao+v51FA4+AVmma+l9wUTUHVM2T5NivUuN2aGg4VzBVym665MuZTjHKGQVhSCiDBXuaP6LBrd9X0==JLIM"}
```

I guess I'm trying to do something similar to what is described in #965 but I do not get any error messages.

**Comment by comotion** [comment]
oh, and `camget -shared=theshareurl -o foo` yeilds an empty directory foo/shared-20180214111614/

Likewise, selecting two items and hitting "share" directly gives a similar json output when following the share link (while I was expecting an ui gallery..)  However, in this case, camget does dump the shared files.

**Comment by mpl** [comment]
There are two different points here:

-sharing of dynamic sets is not (yet) supported, only of files and directories. I suppose we could make camget transform a set into a directory on the fly, as we do for the "Download" action. I don't know/remember why I haven't done it yet. Sets and directories aren't equivalent, but since we already do it for Download, I suppose it might be fair to do it for camget. I'll open a different issue to at least start a discussion about it.

-a share of multiple items has no "magic" representation in the web UI. The reason a shared URL of a single item "shows" the underlying shared item in the web UI, is because the share handler has this special "assemble" parameter that asks the handler to fetch the shared item and send it back as a file. The behavior is trivial to decide for a single item, but it's not the case for several items. What do you send back then? you want the web UI to show you a gallery, but what if the client is a command-line client? Then it makes no sense to reply with a bunch of html, and maybe a zip of all the items makes more sense. etc.
Long story short, I do think it is an important feature, and I do think as you do, that the web UI should have some sort of user-friendly representation for shares of multiple items, but @bradfitz disagrees that it's a priority. So for now it's postponed at best.

-lastly, I don't really know what  #965 is about, since the guy never bothered replying with more detail.

**Comment by mpl** [comment]
hmm, as opposed to what I said in the first point, we apparently do not support Downloading of sets either, my bad.
The 'camtool makestatic' command line client transforms a set into a directory (that you can then Download or Share), and maybe the web UI should do that somehow under the covers as well.

**Comment by mpl** [comment]
I've just opened #1058 to discuss point 1.

**Comment by comotion** [comment]
ok, I think I understand the hows and whys, however I'd still like to be able to share a gallery of pictures. Even with 3 or 5 steps. Is it possible today, perhaps througha static directory?

**Comment by mpl** [comment]
In the general sense of the term, there currently are two ways you can "share" a set of pictures, and have the recipient view them as a gallery:

1) If they also have a Perkeep server:
  a) transform your set into a static dir with 'camtool makestatic'
  b) create a share claim for the newly created dir (either with Share from the web UI, or with camput share)
  c) give them the share link. They can can then import the images with camget, and camput (file -filenodes) them as a set into their perkeep.
  c bis) or, once https://camlistore-review.googlesource.com/c/camlistore/+/10006 lands they can directly import with one click your shared blobs into their Perkeep with their web UI.

2) If they don't have a Perkeep server:
  a) Set up the publisher app
  b) create a camliPath for the permanode of your pictures, on the camliRoot node for the publisher
  c) Share the published path to whoever (it will look like that: https://granivo.re/pics/thabor-20160912)

**Comment by comotion** [comment]
Ah, yes, the publisher :-)

So I've followed the guide here https://github.com/perkeep/perkeep/tree/master/doc/publishing
to set up the publisher with
```
    "publish": {
        "/pics/": {
            "camliRoot": "pics",
            "cacheRoot": "/home/comotion/var/camlistore/blobs/cache",
            "goTemplate": "gallery.html"
        }
    }

```
And I see it starting:
```
PUBLISHER: 2018/02/20 14:12:28 Starting publisher version 2017-12-30-1d4d49ad7a; Go go1.9.3 (linux/amd64)
PUBLISHER: 2018/02/20 14:12:28 Starting to listen on http://localhost:44701
```

Since the UI doesn't allow the creation of permanodes, I camput a dummy permanode with camliRoot = pics and camliPath:images = sha1-<images>
the sha I point to is a permanode with some camliContent. 
I get a 403 error on http://cereal.delta9.pl:3179/pics/ and a 404 error on http://cereal.delta9.pl:3179/pics/images

I suppose the publisher works on a different port, but I get a 403 error at
http://cereal.delta9.pl:3155 and 404 errors on http://cereal.delta9.pl:3155/pics and http://cereal.delta9.pl:3155/pics/images

however,
http://cereal.delta9.pl:3155/images does not present a gallery, but does provide the `some pics.zip`

My goal is still just a gallery of images served from perkeep. Where did I go wrong?

**Comment by mpl** [comment]
I've opened #1065 to deal directly with the last point you reported, which is indeed a bug.

Otherwise, I believe I've answered your other questions, so I'm closing this. Feel free to reopen if i forgot something.

**Comment by comotion** [comment]
Yes you did answer my questions, thank you @mpl. Looking forward to a fix for #1065.

**Comment by mpl** [comment]
@comotion you can try https://camlistore-review.googlesource.com/c/camlistore/+/14486 out yourself until it gets merged if you want. I'm pretty confident about it, and it's pretty harmless as it only affects the publisher.
